accounts-db/src/account_locks.rs

```diff
@@ -21,10 +21,61 @@ impl AccountLocks {
     /// The bool in the tuple indicates if the account is writable.
     /// Returns an error if any of the accounts are already locked in a way
     /// that conflicts with the requested lock.
+    /// NOTE this is the pre-SIMD83 logic and can be removed once SIMD83 is active.
     pub fn try_lock_accounts<'a>(
         &mut self,
         keys: impl Iterator<Item = (&'a Pubkey, bool)> + Clone,
-    ) -> Result<(), TransactionError> {
+    ) -> Result<()> {
+        self.can_lock_accounts(keys.clone())?;
+        self.lock_accounts(keys);
+
+        Ok(())
+    }
+
+    /// Lock accounts for all transactions in a batch which don't conflict
+    /// with existing locks. Returns a vector of `TransactionResult` indicating
+    /// success or failure for each transaction in the batch.
+    /// NOTE this is the SIMD83 logic; after the feature is active, it becomes
+    /// the only logic, and this note can be removed with the feature gate.
+    pub fn try_lock_transaction_batch<'a>(
+        &mut self,
+        validated_batch_keys: impl Iterator<
+            Item = Result<impl Iterator<Item = (&'a Pubkey, bool)> + Clone>,
+        >,
+    ) -> Vec<Result<()>> {
+        let available_batch_keys: Vec<_> = validated_batch_keys
+            .map(|validated_keys| {
+                validated_keys
+                    .clone()
+                    .and_then(|keys| self.can_lock_accounts(keys))
+                    .and(validated_keys)
```

Comment:
what does this `.and` do?

Reply:
its the same as `.and_then(|_| validated_keys)`, i wrote `can_lock_accounts()` to have the same types as all the similar existing method in this struct, so it accepts a keys iterator and returns `Result<()>`

at this stage of validation we have a collection of passes and fails, the `.and_then()` takes only the passes and might fail some more, the `.and()` turns the passes back into the keys they should have. we could make `can_lock_accounts()` clone and return the `impl Iter` which would basically be the same thing